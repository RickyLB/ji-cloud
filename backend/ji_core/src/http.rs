use std::net::{SocketAddr, TcpListener};

/// Creates a [`TcpListener`] via `listenfd` if possible.
///
/// [`TcpListener`]: https://doc.rust-lang.org/stable/std/net/struct.TcpListener.html
pub fn get_tcp_fd() -> Option<TcpListener> {
    #[cfg(feature = "listenfd")]
    let fd = listenfd::ListenFd::from_env().take_tcp_listener(0).unwrap();

    #[cfg(not(feature = "listenfd"))]
    let fd = None;

    fd
}

/// Get the port to run the server on.
pub fn get_addr(default: Option<u16>) -> SocketAddr {
    // 0 is a special port that means "assign me whatever port you want"
    let port = std::env::var("PORT")
        .ok()
        .and_then(|it| it.parse().ok())
        .or(default)
        .unwrap_or(0);

    ([0, 0, 0, 0], port).into()
}
