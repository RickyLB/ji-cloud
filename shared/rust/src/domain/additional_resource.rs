//! Types for additional resources for JIG or Courses.

use crate::domain::{
    asset::AssetId, audio::AudioId, image::ImageId, meta::ResourceTypeId, pdf::PdfId,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Wrapper type around [`Uuid`](Uuid), represents the ID of an additional resource.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
pub struct AdditionalResourceId(pub Uuid);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
/// Over-the-wire representation of a JIG or Course additional resource.
pub struct AdditionalResource {
    /// The additional resources's ID.
    pub id: AdditionalResourceId,

    /// Name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Content of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

/// Request to create a new `AdditionalResource`.
///
/// [`additional_resource::Create`](crate::api::endpoints::additional_resource::Create)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceCreateRequest {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    pub asset_id: AssetId,

    /// Display name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Value of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

/// Request to update an `AdditionalResource`.
///
/// [`additional_resource::Update`](crate::api::endpoints::additional_resource::Update)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceUpdateRequest {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,

    /// resource display name
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// Type of additional  resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub resource_type_id: Option<ResourceTypeId>,

    /// Kind of additional resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(flatten)]
    pub resource_content: Option<ResourceContent>,
}

/// Delete an `AdditionalResource` by Asset Id.
///
/// [`additional_resource::GetLive`](crate::api::endpoints::additional_resource::GetLive)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AssetIdResource {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,
}

/// Value of additional resource
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ResourceContent {
    /// Additional resource kind: image
    ImageId(ImageId),
    /// Additional resource kind: audioFile
    AudioId(AudioId),
    /// Additional resource kind: link
    Link(url::Url),
    /// Additional resource kind: pdf
    PdfId(PdfId),
}

into_uuid![AdditionalResourceId];
