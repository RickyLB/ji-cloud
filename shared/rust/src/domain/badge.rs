//! Types for Badges.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::asset::UserOrMe;

/// Wrapper type around [`Uuid`], represents the ID of a Badge.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
pub struct BadgeId(pub Uuid);

/// The response returned when a request for `GET`ing a Badge is successful.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Badge {
    /// The ID of the Badge.
    pub id: BadgeId,

    /// The Badge's name.
    pub display_name: String,

    /// Creator of Badge
    pub created_by: Uuid,

    /// Description of the Badge. Defaults to empty string.
    pub description: String,

    /// Number of members on Badge
    pub member_count: u32,

    /// Thumbnail URL
    pub thumbnail: url::Url,

    /// When Badge was created
    pub created_at: DateTime<Utc>,

    /// When Badge was last edited
    pub last_edited: Option<DateTime<Utc>>,
}

/// Request to create a new Badge.
///
/// This creates the draft and live [Badge Data](Badge Data) copies with the requested info.
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BadgeCreateRequest {
    /// The Badge's name.
    pub display_name: String,

    /// Description of the Badge. Defaults to empty string.
    pub description: String,

    /// Thumbnail of the Badge
    pub thumbnail: url::Url,
}

/// Request for updating a Badge's draft data.
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct BadgeUpdateRequest {
    /// The Badge's name to be updated.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// The current author to be updated
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub creator_id: Option<Uuid>,

    /// Description of the Badge to be updated.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub description: Option<String>,

    /// Thumbnail of the Badge to be updated.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub thumbnail: Option<url::Url>,
}

/// Query for [`Browse`](crate::api::endpoints::badge::Browse).
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct BadgeBrowseQuery {
    /// creator of badges
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_id: Option<UserOrMe>,

    /// The page number to get.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page: Option<u32>,

    /// The hits per page to be returned
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_limit: Option<u32>,
}

/// Response for [`Browse`](crate::api::endpoints::badge::Browse).
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BadgeBrowseResponse {
    /// the Badges returned.
    pub badges: Vec<Badge>,

    /// The number of pages found.
    pub pages: u32,

    /// The total number of Badges found
    pub total_badge_count: u64,
}

/// Search for Badges via the given query string.
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct BadgeSearchQuery {
    /// The query string.
    #[serde(default)]
    #[serde(skip_serializing_if = "String::is_empty")]
    pub q: String,

    /// The page number of the Badges to get.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page: Option<u32>,

    /// Optionally filter by author's id
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_id: Option<UserOrMe>,

    /// Optionally filter by the author's name
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_name: Option<String>,

    /// The hits per page to be returned
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_limit: Option<u32>,
}

/// Response for [`Search`](crate::api::endpoints::badge::Search).
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BadgeSearchResponse {
    /// the Badges returned.
    pub badges: Vec<Badge>,

    /// The number of pages found.
    pub pages: u32,

    /// The total number of Badges found
    pub total_badge_count: u64,
}

/// Members associated with Badge
#[derive(Serialize, Deserialize)]
pub struct BrowseMembersResponse {
    /// user id of member
    pub members: Vec<Uuid>,

    /// user id of member
    pub count: u32,
}

into_uuid![BadgeId];
