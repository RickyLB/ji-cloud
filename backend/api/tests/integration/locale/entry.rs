use http::StatusCode;

use macros::test_service;
use serde_json::json;
use shared::domain::locale::{CreateEntryRequest, EntryStatus};
use sqlx::postgres::{PgConnectOptions, PgPoolOptions};

use crate::{
    fixture::Fixture,
    helpers::{setup_service, LoginExt},
};

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn delete(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .delete(&format!("http://0.0.0.0:{}/v1/locale/entry/3", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    Ok(())
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn get(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/locale/entry/3", port))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

async fn list(query: &[(&str, &str)], port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/locale/entry", port))
        .query(query)
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_all_by_default(port: u16) -> anyhow::Result<()> {
    list(&[], port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_all_by_bundle(port: u16) -> anyhow::Result<()> {
    list(&[("groupBy", "bundle")], port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_empty_bundle_by_default(port: u16) -> anyhow::Result<()> {
    list(&[("bundles", "85a46ffe-7c67-11eb-a0d7-277d94fe130c")], port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_empty_bundle_by_bundle(port: u16) -> anyhow::Result<()> {
    list(
        &[
            ("groupBy", "bundle"),
            ("bundles", "85a46ffe-7c67-11eb-a0d7-277d94fe130c"),
        ],
        port,
    )
    .await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_single_bundle_by_default(port: u16) -> anyhow::Result<()> {
    list(&[("bundles", "8359a48a-7c67-11eb-a0d7-0fd74777a62c")], port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_single_bundle_by_bundle(port: u16) -> anyhow::Result<()> {
    list(
        &[
            ("groupBy", "bundle"),
            ("bundles", "8359a48a-7c67-11eb-a0d7-0fd74777a62c"),
        ],
        port,
    )
    .await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn create(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/locale/entry", port))
        .json(&CreateEntryRequest {
            bundle_id: "8359a48a-7c67-11eb-a0d7-0fd74777a62c".parse()?,
            section: None,
            item_kind_id: None,
            english: Some("Just another test entry".to_owned()),
            hebrew: None,
            status: EntryStatus::Discuss,
            zeplin_reference: None,
            comments: None,
            in_app: false,
            in_element: true,
            in_mock: true,
        })
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".id" => "[id]"});

    Ok(())
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn update_in_app(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!("http://0.0.0.0:{}/v1/locale/entry/3", port))
        .json(&json! ({
            "inApp": false,
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/locale/entry", port))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}
