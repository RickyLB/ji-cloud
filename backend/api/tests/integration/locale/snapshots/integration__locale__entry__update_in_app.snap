---
source: tests/integration/locale/entry.rs
expression: body
---
{
  "list": [
    {
      "id": 1,
      "bundleId": "7c11d04e-7c67-11eb-a0d7-9b5fb8f29de9",
      "section": "A",
      "itemKindId": "1623eab4-7c68-11eb-a0d7-5f27277abc1b",
      "english": "Entry #1",
      "hebrew": null,
      "status": "approved",
      "zeplinReference": null,
      "comments": null,
      "inApp": true,
      "inElement": true,
      "inMock": false
    },
    {
      "id": 2,
      "bundleId": "8359a48a-7c67-11eb-a0d7-0fd74777a62c",
      "section": "A",
      "itemKindId": "1623eab4-7c68-11eb-a0d7-5f27277abc1b",
      "english": "Entry #2",
      "hebrew": null,
      "status": "approved",
      "zeplinReference": null,
      "comments": null,
      "inApp": true,
      "inElement": true,
      "inMock": false
    },
    {
      "id": 3,
      "bundleId": "7c11d04e-7c67-11eb-a0d7-9b5fb8f29de9",
      "section": "A",
      "itemKindId": "1623eab4-7c68-11eb-a0d7-5f27277abc1b",
      "english": "Entry #3",
      "hebrew": null,
      "status": "approved",
      "zeplinReference": null,
      "comments": null,
      "inApp": false,
      "inElement": true,
      "inMock": false
    },
    {
      "id": 4,
      "bundleId": "7c11d04e-7c67-11eb-a0d7-9b5fb8f29de9",
      "section": "B",
      "itemKindId": "1623eab4-7c68-11eb-a0d7-5f27277abc1b",
      "english": "Entry #B-1",
      "hebrew": null,
      "status": "approved",
      "zeplinReference": null,
      "comments": null,
      "inApp": true,
      "inElement": true,
      "inMock": false
    },
    {
      "id": 5,
      "bundleId": "7c11d04e-7c67-11eb-a0d7-9b5fb8f29de9",
      "section": "C",
      "itemKindId": "1623eab4-7c68-11eb-a0d7-5f27277abc1b",
      "english": "Entry #B-1",
      "hebrew": null,
      "status": "approved",
      "zeplinReference": null,
      "comments": null,
      "inApp": true,
      "inElement": true,
      "inMock": false
    }
  ]
}
